src/metabase/models/card.clj

```diff
@@ -473,6 +488,292 @@
+  "Removes all alerts and will email each recipient letting them know"
+  [alerts]
+  (delete-alert-and-notify! messages/send-alert-stopped-because-archived-email! alerts))
+
+(defn- delete-alert-and-notify-changed! [alerts]
+  (delete-alert-and-notify! messages/send-alert-stopped-because-changed-email! alerts))
+
+(defn- delete-alerts-if-needed! [old-card {card-id :id :as new-card}]
+  ;; If there are alerts, we need to check to ensure the card change doesn't invalidate the alert
+  (when-let [alerts (seq (pulse/retrieve-alerts-for-cards {:card-ids [card-id]}))]
+    (cond
+
+      (card-archived? old-card new-card)
+      (delete-alert-and-notify-archived! alerts)
+
+      (or (display-change-broke-alert? old-card new-card)
+          (goal-missing? old-card new-card)
+          (multiple-breakouts? new-card))
+      (delete-alert-and-notify-changed! alerts)
+
+      ;; The change doesn't invalidate the alert, do nothing
+      :else
+      nil)))
+
+(defn- card-is-verified?
+  "Return true if card is verified, false otherwise. Assumes that moderation reviews are ordered so that the most recent
+  is the first. This is the case from the hydration function for moderation_reviews."
+  [card]
+  (-> card :moderation_reviews first :status #{"verified"} boolean))
+
+(defn- changed?
+  "Return whether there were any changes in the objects at the keys for `consider`.
+
+  returns false because changes to collection_id are ignored:
+  (changed? #{:description}
+            {:collection_id 1 :description \"foo\"}
+            {:collection_id 2 :description \"foo\"})
+
+  returns true:
+  (changed? #{:description}
+            {:collection_id 1 :description \"foo\"}
+            {:collection_id 2 :description \"diff\"})"
+  [consider card-before updates]
+  ;; have to ignore keyword vs strings over api. `{:type :query}` vs `{:type "query"}`
+  (let [prepare              (fn prepare [card] (walk/prewalk (fn [x] (if (keyword? x)
+                                                                        (name x)
+                                                                        x))
+                                                              card))
+        before               (prepare (select-keys card-before consider))
+        after                (prepare (select-keys updates consider))
+        [_ changes-in-after] (data/diff before after)]
+    (boolean (seq changes-in-after))))
+
+(def ^:private card-compare-keys
+  "When comparing a card to possibly unverify, only consider these keys as changing something 'important' about the
+  query."
+  #{:table_id
+    :database_id
+    :query_type ;; these first three may not even be changeable
+    :dataset_query})
+
+(defn update-card!
+  "Update a Card. Metadata is fetched asynchronously. If it is ready before [[metadata-sync-wait-ms]] elapses it will be
+  included, otherwise the metadata will be saved to the database asynchronously."
+  [{:keys [id] :as card-before-update} card-updates]
```

Comment:
wdyt about adding `actor-id` as part of this function's parameters? it'll make this function free of dynamic vars

Reply:
Yes, good idea [Ngoc's suggestions](https://github.com/metabase/metabase/pull/36123/commits/03ff937ed888e550c23d6e6bf2aa1ee9c7e5d6b7)